firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/serialization/FirestoreMapDecoder.kt

```diff
@@ -270,11 +308,43 @@ private class FirestoreListDecoder(
      */
     override fun decodeElementIndex(descriptor: SerialDescriptor): Int {
         if (elementIndex == decodeValueList.size) return CompositeDecoder.DECODE_DONE
-        currentDecodeElement = Element(decodeValueList[elementIndex], elementIndex)
+        currentDecodeElement = Element(decodeValueList[elementIndex], elementIndex, null)
         return elementIndex++
     }
 }
 
+/**
+ * A data class that stores the information of all the annotations (both Java and Ktx) associated
+ * with each field name, and the Java class types associated with each field name.
+ */
+class JavaAndKtxAnnotationMapper(clazz: Class<*>) {
```

Comment:
This is the main logic of reflection, I have to use Kotlin reflection to guide Java reflection (java reflection cannot get all the fields from super class, while Kotlin has this native method).

Also need Kotlin reflection to get the @SerialName, so that I can match the java annotations' owner with the elements in `descriptor`.

Need to handle `List<CustomClass>` with some Kotlin trick, Java will erase generic type, While Kotlin can obtain this from `KType` (not `KClass`)

BUT, I am not 100% sure if this method is possible to support recursive generic type or not.